### ISSUE TITLE
ignoring_interference_by_writer for validate_numericality_of

### ISSUE THREAD

**Issue by ArturT** [thoughtbot/shoulda-matchers#799]
I'm using shoulda-matchers 3.0.0. It would be cool if `validate_numericality_of` could take `ignoring_interference_by_writer`.

I have validator class and it runs validation on ModelEntity which is created by virtus gem. Virtus does type casting to keep my field as Integer and that's why validate_numericality_of fails.

```
Failure/Error: it { should validate_numericality_of(:node_total).is_greater_than_or_equal_to(2) }
     Shoulda::Matchers::ActiveModel::AllowValueMatcher::CouldNotSetAttributeError:
       Expected Class to be able to set node_total to "1", but got 1 instead.
```

**Comment by mcmire** [comment]
Ah. That's a good point. If you have a plain Ruby class, there's no way for `validate_numericality_of` to know what to do.

I hesitate to add this qualifier since it means that people could blindly apply it. I'm wondering if I could add an `cast_to` qualifier that lets you specify the type of values that the attribute hold. So in this case you might say:

``` ruby
it { should validate_numericality_of(:node_total).cast_to(Integer).is_greater_than_or_equal_to(2) }
```

Essentially, what this qualifier would do is change the matcher so that if it detects a mismatch in values, it will use the class you've given to typecast the value being set and compare it to the value read back. So it'd say `Integer("1") == 1` and since that's true it would make the matcher pass.

I'll work on that next.

**Comment by ArturT** [comment]
@mcmire Sounds good. Thanks!

**Comment by jetaggart** [comment]
I'm seeing a similar issue with rails 4.2.4. It seems their casting logic has changed and a string gets casted to a numeric based on the column (I'm failing to find this in the changelog, so I could be mistaken). For example:

```
Failure/Error: it { is_expected.to validate_numericality_of(:order).is_greater_than_or_equal_to(0) }
     Shoulda::Matchers::ActiveModel::AllowValueMatcher::CouldNotSetAttributeError:
       Expected Class to be able to set order to "abcd", but got 0 instead.
```

**Comment by mcmire** [comment]
@jetaggart Sorry Jeff. Yup, you're right that it isn't in the changelog, but it's directly due to the work on all the new typecasting stuff that @sgrif did in Rails 4.2.

**Comment by jetaggart** [comment]
It's a good change, even if @sgrif is making my life difficult (:heart:). I'd be happy to help get things working with the new casting. What do you think the best solution is? Just produce values for that type?

**Comment by mcmire** [comment]
Well, there are actually two things happening here:

* The `"abcd"` value you're seeing in the error message is actually being provided by the matcher itself. Right now the `validate_numericality_of` matcher assumes that your column is a string, so it tries to set the attribute to a string. Prior to Rails 4.2 this wouldn't be a problem, but due to the new typecasting behavior, that string will get converted to an integer.
* The `allow_value` matcher (which `validate_numericality_of` uses internally) is confused because the value going into the attribute isn't the value coming out. Prior to shoulda-matchers 3.0, this wouldn't be a problem, but now a CouldNotSetAttributeError is raised if this happens.

In short, the bug is in shoulda-matchers itself. There's already a fix in place so that our numericality matcher doesn't use `"abcd"` anymore if your column is numeric, but that wouldn't exactly make your test pass. So the bad news is that there isn't anything you can do about it just yet. There's a discussion happening in #800 for how to solve this problem, so if you have any input on that, then it would be appreciated, but otherwise, hang tight and I'll try to put out a release shortly. Sorry :/

**Comment by jetaggart** [comment]
No worries, thanks for letting me know. I'll check out that PR.

**Comment by bdewater** [comment]
@jetaggart slightly offtopic but I was bitten by that change as well on a project, I believe this is the commit: https://github.com/rails/rails/commit/d8e710410ea300ec4626250c0b35946cb52bc38c

**Comment by mcmire** [comment]
@ArturT @jetaggart @bdewater There's now a fix for this in master. Would you be willing to point to the GitHub repo in your Gemfile and let me know if that resolves this issue?

**Comment by mcmire** [comment]
3.1.0 is now released with the aforementioned fix.

**Comment by ArturT** [comment]
@mcmire Thanks for fix. It works now. 

I had tiny problem because I'm testing `BuildValidator` class instead of active model. `BuildValidator` was using `method_missing` to pass call to underling entity (I'm using virtus gem). Method `attribute_exists?` in https://github.com/thoughtbot/shoulda-matchers/commit/296211211497e624dde87adae68b385ad4cdae3a/lib/shoulda/matchers/active_model/allow_value_matcher/attribute_setter.rb#diff-4 uses `respond_to?` and because of that I saw error `Shoulda::Matchers::ActiveModel::AllowValueMatcher::AttributeDoesNotExistError`. Then I found out that I don't have `respond_to_missing?` in `BuildValidator` class so I added it which actually helped. 

Hope this tip will be helpful if someone will have the same problem.

**Comment by jetaggart** [comment]
@mcmire sorry for the delay, this is working for our simple cases. Thanks for the fix!

**Comment by mcmire** [comment]
@jetaggart Awesome! Glad to hear.
